File: 109.有序链表转换二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=109 lang=cpp
 *
 * [109] 有序链表转换二叉搜索树
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    // ListNode *getMedian(ListNode *left, ListNode *right)
    // {
    //     ListNode *fast = left;
    //     ListNode *slow = left;
    //     while (fast != right && fast->next != right)
    //     {
    //         fast = fast->next->next;
    //         slow = slow->next;
    //     }
    //     return slow;
    // }
    int getLength(ListNode *head)
    {
        int length = 0;
        for (; head != nullptr; head = head->next, length++)
            ;
        return length;
    }
    TreeNode *buildTree(ListNode *&head, int left, int right)
    {
        if (left > right)
            return nullptr;
        int mid = (left + right) / 2;
        TreeNode *root = new TreeNode();
        root->left = buildTree(head, left, mid - 1);
        root->val = head->val;
        head = head->next;
        root->right = buildTree(head, mid + 1, right);
        return root;
    }
    TreeNode *sortedListToBST(ListNode *head)
    {
        int length = getLength(head);
        return buildTree(head, 0, length - 1);
    }
};
// @lc code=end

File: 109.有序链表转换二叉搜索树.py
#
# @lc app=leetcode.cn id=109 lang=python3
#
# [109] 有序链表转换二叉搜索树
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # def getMedian(self, left: ListNode, right: ListNode) -> ListNode:
    #     fast, slow = left, left
    #     while fast != right and fast.next != right:
    #         fast = fast.next.next
    #         slow = slow.next
    #     return slow

    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def getLength(head: ListNode) -> int:
            length = 0
            while head:
                head = head.next
                length += 1
            return length

        def buildTree(left: int, right: int) -> TreeNode:
            if left > right:
                return None
            mid = (left+right)//2
            root = TreeNode()
            root.left = buildTree(left, mid-1)
            nonlocal head
            root.val = head.val
            head = head.next
            root.right = buildTree(mid+1, right)
            return root

        length = getLength(head)
        return buildTree(0, length-1)
# @lc code=end
